### PR TITLE
feature/pid-5-manage-handle-table

### LINKED ISSUE #64: feature/pid-5-manage-handle-table
**ETA:** 20.8

## Problem description
The table should have this functions: [handle table clarin](https://github.com/ufal/clarin-dspace/wiki/PIDs-Persistent-Identifiers#new-gui-for-managing-handles)

Create a new option 'handles' in the left side menu - this option will see only admin.
After clicking to 'handles' option the admin will be redirected to the handle page where will be a table with handles. Handle should provide operation: 'update handle', 'change handle prefix',.. - the same as in the clarin wiki

### How to test 
**New external handle**
- [ ] If you click  to the button `New external handle` you will be redirected to the New handle page
- [ ] The new handle page has two input fields: `Handle` and `URL`
- [ ] If you click Submit it will redirect you to the page with table
**Edit handle**
**Delete handle**
**Change global prefix**

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|-----:|-----:|-------:|
| ETA                  |  20.8  |    0 |     0 |      0 |        0 |
| Developing      |  2+7+8.5+8+7+4+2.5+6+4  |    0 |    0 |      0 |         0 |
| Review             |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -   |   -     |        0 |
## Problem description
The table should have this functions: [handle table clarin](https://github.com/ufal/clarin-dspace/wiki/PIDs-Persistent-Identifiers#new-gui-for-managing-handles)

Create a new option 'handles' in the left side menu - this option will see only admin.
After clicking to 'handles' option the admin will be redirected to the handle page where will be a table with handles. Handle should provide operation: 'update handle', 'change handle prefix',.. - the same as in the clarin wiki
- [x] Create a new handle page '/handles'. Class `admin-sidebar.component.ts`
- [x] Create Test for a new handle option in the menu
- [x] Create a redirection to the '/handle' page after click on the menu option
- [x] Create a new Handle page '/handle'
- [x] Create a new component - handle table for managing handles
- [x] Create messages for handle menu and handle page in the `en.json5`
- [x] Create `handle.model.ts` object - inspire by `metadata-value.model.ts`
- [x] Create a handle service to send/retrieve requests from BE
- [x] Create option to change handle prefix
- [x] Create option to change whole handle
- [x] Add URL to the handle object
- [x] Finish everything when the BE will be fixed
- [x] Create tests:
- [x] 1. Handle option is in the admin menu
- [x] 2. IT - should redirect to the Handle page
- [x] 3. Handle data service - there is no need to create test because I didn't create any new method
- [x] 4. change-handle-prefix-page.component.ts
- [x] 5. edit-handle page component
- [x] 6. handle global actions component - there is no need to create test because this component is very simple
- [x] 8. handle table component
- [x] 9. new handle page component
- [x] Sort handles
- [x] Change handle prefix - dropdown of all prefixes

## Analysis
(Write here, if there is needed describe some specific problem. Erase it, when it is not needed.)
### 1. Fetching Handles data to the table takes too long
**Reason:** Converting DSpaceObject from the BE takes too long
**Solution:** Removed DSpaceObject attribute from Handle object

### 2. The table data wasn't updated properly after changing pagination
**Reason:** when the table page was changed the component `<ds-pagination>` do not uploads _options_ attribute before sending request to the BE. That means FE sends request with old pagination options.
**Solution:** send the request only if the options has changed or is Init

### 3. After deleting the the handle the table should be refreshed with fresh data but it didn't happened because the table still loads deleted handle from BE.
**Reason:** After delete handle request is called refresh table request immediately but in the BE the handle is not deleted while refresh table request is called so the refresh table request still loads deleted handle.
**Solution:** Refresh the table only if is changed the table data in the BE.

### 4. Pagination error
**Problem:**
When I edited some Handle I was redirected to the handle table page, and when I click on the next page it redirects me to the editing handle page again.
**Reason:**
The redirecting was doing in the subscription and that subscription was still active.
**Solution:**
Use `take(1)` method which take just the first value from the subscription and end it.
### 5. TranslateService Error
**Problem:** when I provided TranslateService which was defined like `jasmine.createSpyObj(..)` I cannot run the tests because of ` this.translate.get(...).subscribe is not a function` error. 
**Reason:** I do not know right reason but when I removed `fixture.detectChanges();` from `beforeEach` the error disapeared.
**Solution:** Remove  `fixture.detectChanges();` from the `beforeEach` .
**Reason:**

### CODE REVIEW

File: src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<div class="container">
```
Comment (mrovnanik) [inline src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html:1]: why the empty file? if it's necessary, leave a comment or something to make it clear why it's empty
https://github.com/dataquest-dev/dspace-angular/pull/67/files#diff-39df97adcc4c9db508881d05bca923399a8752717286c594b4abee38b10a2a8e

File: src/assets/i18n/en.json5 [hunk 2]
```diff
@@ -1571,6 +1571,112 @@
...
+  "handle-table.new-handle.form-handle-input-placeholder": "Enter handle",
+
+  "handle-table.new-handle.form-url-input-text": "URL",
+
+  "handle-table.new-handle.form-url-input-placeholder": "Enter URL",
+
+  "handle-table.new-handle.form-button-submit": "Submit",
+
+
+  "handle-table.new-handle.notify.error": "Server Error - Cannot create a new handle",
```
Comment (mrovnanik) [inline src/assets/i18n/en.json5]: Cannot create new handle. No need for 'a'. Consider removing 'a' also in the other instances

File: src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html (hunk 1 above)
Comment (mrovnanik) [reply src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html:1]: consider adding the same comment to all other instances of empty files

**Review by mrovnanik — CHANGES_REQUESTED** [review]

File: src/assets/i18n/en.json5 (hunk 2 above)
Comment (milanmajchrak) [reply src/assets/i18n/en.json5]: Removed 'a'

File: src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html (hunk 1 above)
Comment (milanmajchrak) [reply src/app/handle-page/change-handle-prefix-page/change-handle-prefix-page.component.html:1]: Added comments to the *.scss files that they are styling files.

**Review by mrovnanik — APPROVED** [review]

**Pull request merged**